File: 2023/16/16.rb
#!/usr/bin/env ruby

grid = []
$stdin.readlines.lazy.map(&:chomp).each do |line|
  grid << line.chars
end

def count_energized_positions(grid, start)
  energized_positions = {}
  visited = {}
  positions = [start]

  y_range = 0...grid.size
  x_range = 0...grid.first.size

  while positions.any?
    visit = positions.shift
    (px, py), (dx, dy) = visit
    x, y = px + dx, py + dy
    next unless x_range.include?(x) and y_range.include?(y) # Exited the grid
    next if visited[visit]
    visited[visit] = true
    energized_positions[[x, y]] = true

    case grid[y][x]
    when '.'
      positions << [[x, y], [dx, dy]]
    when '/'
      if dx == 0
        ndx = dy > 0 ? -1 : 1
        ndy = 0
      else
        ndx = 0
        ndy = dx > 0 ? -1 : 1
      end
      positions << [[x, y], [ndx, ndy]]
    when '\\'
      if dx == 0
        ndx = dy > 0 ? 1 : -1
        ndy = 0
      else
        ndx = 0
        ndy = dx > 0 ? 1 : -1
      end
      positions << [[x, y], [ndx, ndy]]
    when '-'
      if dx == 0
        positions << [[x, y], [-1, 0]]
        positions << [[x, y], [1, 0]]
      else
        positions << [[x, y], [dx, dy]]
      end
    when '|'
      if dy == 0
        positions << [[x, y], [0, -1]]
        positions << [[x, y], [0, 1]]
      else
        positions << [[x, y], [dx, dy]]
      end
    end
  end
  energized_positions.size
end

# Part 1
puts count_energized_positions(grid, [[-1, 0], [1, 0]])

# Part 2
most_energized = 0

# left and right edges
(0...grid.size).each do |y|
  most_energized = [
    most_energized,
    count_energized_positions(grid, [[-1, y], [1, 0]]),
    count_energized_positions(grid, [[grid.first.size, y], [-1, 0]]),
  ].max
end

# top and bottom edges
(0...grid.first.size).each do |x|
  most_energized = [
    most_energized,
    count_energized_positions(grid, [[x, -1], [0, 1]]),
    count_energized_positions(grid, [[x, grid.size], [0, -1]]),
  ].max
end

puts most_energized

File: 2023/16/16.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type V struct {
	x  int
	y  int
	dx int
	dy int
}

func countEnergizedPositions(grid []string, start V) int {
	gy, gx := len(grid)-1, len(grid[0])-1
	visited := map[uint64]bool{}
	positions := []V{start}
	energizedPositions := map[uint32]bool{}
	var visit V
	var ndx, ndy int

	for len(positions) > 0 {
		visit, positions = positions[0], positions[1:]
		dx, dy := visit.dx, visit.dy
		x, y := visit.x+dx, visit.y+dy
		if x < 0 || x > gx || y < 0 || y > gy {
			continue
		}

		p := uint64(visit.x) + (uint64(visit.y) << 16) + (uint64(dx) << 32) + (uint64(dy) << 48)
		if visited[p] {
			continue
		}
		visited[p] = true
		energizedPositions[uint32(x)+(uint32(y)<<16)] = true

		switch grid[y][x] {
		case '.':
			positions = append(positions, V{x, y, dx, dy})
		case '/':
			if dx == 0 {
				ndx, ndy = -dy, 0
			} else {
				ndx, ndy = 0, -dx
			}
			positions = append(positions, V{x, y, ndx, ndy})
		case '\\':
			if dx == 0 {
				ndx, ndy = dy, 0
			} else {
				ndx, ndy = 0, dx
			}
			positions = append(positions, V{x, y, ndx, ndy})
		case '-':
			if dx == 0 {
				positions = append(positions, V{x, y, -1, 0}, V{x, y, 1, 0})
			} else {
				positions = append(positions, V{x, y, dx, dy})
			}
		case '|':
			if dy == 0 {
				positions = append(positions, V{x, y, 0, -1}, V{x, y, 0, 1})
			} else {
				positions = append(positions, V{x, y, dx, dy})
			}
		}
	}
	return len(energizedPositions)
}

func ParseInput() []string {
	scanner := bufio.NewScanner(os.Stdin)
	grid := []string{}
	for scanner.Scan() {
		grid = append(grid, scanner.Text())
	}
	return grid
}

func main() {
	grid := ParseInput()
	fmt.Println(countEnergizedPositions(grid, V{-1, 0, 1, 0}))

	mostEnergized := 0

	for y, r := range grid {
		mostEnergized = max(
			mostEnergized,
			countEnergizedPositions(grid, V{-1, y, 1, 0}),
			countEnergizedPositions(grid, V{len(r), y, -1, 0}),
		)
	}

	for x := range len(grid[0]) {
		mostEnergized = max(
			mostEnergized,
			countEnergizedPositions(grid, V{x, -1, 0, 1}),
			countEnergizedPositions(grid, V{x, len(grid), 0, -1}),
		)
	}

	fmt.Println(mostEnergized)
}
